src/rsa.rs

```diff
@@ -60,61 +60,67 @@ struct RSAVerificationAlgorithm {
+        if try!(decoded.read_byte()) != 0 ||
+           try!(decoded.read_byte()) != 1 {
             return Err(());
         }
-        let decoded = &mut decoded[..signature.len()];
-        try!(bssl::map_result(unsafe {
-            GFp_rsa_public_decrypt(decoded.as_mut_ptr(), decoded.len(),
-                                   n.as_ptr(), n.len(), e.as_ptr(), e.len(),
-                                   signature.as_ptr(), signature.len(),
-                                   self.min_bits, MAX_BITS)
-        }));
-
-        untrusted::Input::from(decoded).read_all((), |decoded| {
-            if try!(decoded.read_byte()) != 0 ||
-               try!(decoded.read_byte()) != 1 {
-                return Err(());
-            }
 
-            let mut ps_len = 0;
-            loop {
-                match try!(decoded.read_byte()) {
-                    0xff => { ps_len += 1; },
-                    0x00 => { break; },
-                    _ => { return Err(()); }
-                }
-            }
-            if ps_len < 8 {
-                return Err(());
+        let mut ps_len = 0;
+        loop {
+            match try!(decoded.read_byte()) {
+                0xff => { ps_len += 1; },
+                0x00 => { break; },
+                _ => { return Err(()); }
             }
+        }
+        if ps_len < 8 {
+            return Err(());
+        }
 
-            let decoded_digestinfo_prefix =
-                try!(decoded.skip_and_get_input(
-                        self.padding_alg.digestinfo_prefix.len()));
-            if decoded_digestinfo_prefix != self.padding_alg.digestinfo_prefix {
-                return Err(());
-            }
+        let decoded_digestinfo_prefix =
+            try!(decoded.skip_and_get_input(padding.digestinfo_prefix.len()));
+        if decoded_digestinfo_prefix != padding.digestinfo_prefix {
+            return Err(());
+        }
 
-            let digest_alg = self.padding_alg.digest_alg;
-            let decoded_digest =
-                try!(decoded.skip_and_get_input(digest_alg.output_len));
-            let digest = digest::digest(digest_alg, msg.as_slice_less_safe());
-            if decoded_digest != digest.as_ref() {
-                return Err(());
-            }
+        let digest_alg = padding.digest_alg;
+        let decoded_digest =
+            try!(decoded.skip_and_get_input(digest_alg.output_len));
+        let digest = digest::digest(digest_alg, msg.as_slice_less_safe());
+        if decoded_digest != digest.as_ref() {
+            return Err(());
+        }
 
-            Ok(())
-        })
+        Ok(())
+    })
+}
+
+
+impl signature_impl::VerificationAlgorithmImpl for RSAVerificationAlgorithm {
+    fn verify(&self, public_key: untrusted::Input, msg: untrusted::Input,
+              signature: untrusted::Input) -> Result<(), ()> {
+        let (n, e) = try!(parse_public_key(public_key));
+        verify(self.padding_alg, msg, signature, (n, e), self.min_bits)
```

Comment:
Please see my comment about adding documentation about the test coverage of `verify_rsa` being dependent on it being called here.

Also, please call it as `signature::primitive::verify_rsa` to make that aspect crystal-clear. (Go ahead and squash all your changes to make that possible, if needed.)